### ISSUE TITLE
What is the difference between this and built-in groups?

### ISSUE THREAD

**Issue by audiolion** [GenePeeks/django-hats#1]
Could you clarify the use case of this over django's contrib.auth groups? If I had a scientist group isn't that the same thing? I can assign permissions to the group, people can be in multiple groups, check groups for permissions, decorate views with required groups, etc.

**Comment by miketheredherring** [comment]
TL;DR - In a nutshell, django-hats is a reusable app, which doesn't try to re-invent the wheel by having Django `Groups` as its foundation. But provides a compact collection of easy to use patterns and utilities for role/permission based checks in templates and CBVs, using clean syntax.

There are a couple of reasons why I decided to make this package, rather than choosing either one of the exiting ones out there, or just vanilla Django `Groups`:
1. Pragmatically defined - Its nice to be able to use the following syntax(see quickstart section of README):
```python
from core.roles import Scientist

def someview(request, *args, **kwargs):
    if Scientist.check_membership(request.user):
        ... Do something cool ...
```

Versus:
```python
from django.contrib.auth.models import Group

def someview(request, *args, **kwargs):
    try:
        scientist_role = Group.objects.get(name='scientist_role')
    except Group.DoesNotExist:
        ... You have to worry about database state ...
    if scientist_role in user.groups.all():
        ... Do something cool ...
```

Where you need to have exception handling for all DB queries, handle initialization/permission assignment for the `Group` yourself, and have a string defined somewhere, or hardcode the name in each occurrence, for manual lookups. Django hats takes care of all of that for you after defining roles and permissions, just import and use.

2. Mixins - Provide and easy way to check for role membership, on top of existing `Permissions`. Roles, like `Groups`, are a great way to assign `Permissions` to a set of `Users`, but in some cases you just really want to check from a specific role to be able to do something rather than a `Permission`.

The closest thing I have used in Django to check `Group` membership on a view would be the `user_passes_test` decorator. First thing here, if you Google 'Django group required', you get a mix of Stackoverflows, Github Gists, and blog posts about how to make your own decorator/mixin, rather than one documented source everyone can use and improve. Second, all these decorators are great for functional based views, but for CBVs a mixin which is easy to use, plays nicely with the built-in `PermissionRequiredMixin`, allows you to mix and match those checks with consistent login/error follow-up behavior.

*Note*: I do realize that `django-braces` has a `GroupRequiredMixin`, but this is another package and falls under the category of the first advantage I pointed out.

3. Templates - Using the provided context processor(see templates section of README), you can set-and-forget and be able to check role membership in templates easily(especially for UI elements which don't tie back directly to a permission):
```html
{% if roles.admin and roles.scientist %}
    ... Show a cool graph ...
{% endif %}
```

Comparatively to querying for the `Group`, and performing the check every view, then adding it to the context, and finally enforcing it in the template.

**Comment by audiolion** [comment]
Hey thanks for this writeup! I would include something like this in the README to discuss the advantages over Groups.

Currently I use groups for memberships, just some comments about my solutions to each of your points:

1. I don't check for database state in code but instead provide a migration with the codebase that checks if the group exists and if not it adds that group to the database

2. I use the `django-braces` package `GroupRequiredMixin` to get the same functionality as you stated

3. I wrote a package called [django-groups-cache](https://github.com/audiolion/django-groups-cache.git) which caches groups so they can be checked without extra database queries on each view. If you don't want to use a cache you can also do

```
# context_processors.py

def groups(request):

    if hasattr(request, 'user'):
        user = request.user
    else:
        # Anonymous User has no Group membership
        return {}
    groups = user.groups.all().values_list('name', flat=True)
    return {
        'groups': groups
    }
```

And add it to context processors, in templates you would do

```
{% if "admin" in groups %}
  ... do something ..
{% endif %}
```


That all being said, I understand the use-case for `django-hats` to bring all those disparate pieces under one roof and provide a nice abstraction. Cool package 👍
